### PR TITLE
SAK-29176 - Allow samigo when publishing a test, to link it to an existing non-externally maintained gradebook item.

### PR DESCRIPTION
**Pull request by fgeraci**
[SAK-29176](https://jira.sakaiproject.org/browse/SAK-29176)

This applies when:

A non-externally maintained GB item exists and can be linked to an assesment when:
    1) Is being published
    2) It was published and has or not, grades already submitted into it
    3) Is retracted and republishes with or without any submissions in it

Should the assessment is disassociated with the GB item, this will be removed.

For retracting the exam, the property "samigo.editPubAssessment.restricted" must be enabled for the code in ActionAuthorListener (~164) defaults to true

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ConfirmPublishAssessmentListener.java [hunk 1]
```diff
@@ -317,12 +319,18 @@ else if (time instanceof java.lang.Boolean) {
...
             getBean("org.sakaiproject.service.gradebook.GradebookExternalAssessmentService");
     }
+    String toGradebook = assessmentSettings.getToDefaultGradebook();
     try{
-	if (assessmentSettings.getToDefaultGradebook() && gbsHelper.isAssignmentDefined(assessmentSettings.getTitle(), g)){
-        String gbConflict_err= ContextUtil.getLocalizedString("org.sakaiproject.tool.assessment.bundle.AssessmentSettingsMessages" , "gbConflict_error");
-        context.addMessage(null,new FacesMessage(gbConflict_err));
-        error=true;
-      }
+    	if (toGradebook!=null && StringUtils.equals(toGradebook, toDefaultGradebook) &&
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ConfirmPublishAssessmentListener.java]: I don't think you need to test for `null` here, as `StringUtils.equals()` is null safe. Same on L:329

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/PublishAssessmentListener.java [hunk 2]
```diff
@@ -267,7 +269,7 @@ private boolean checkTitle(AssessmentFacade assessment){
     }
     String toGradebook = assessment.getEvaluationModel().getToGradeBook();
     try{
-      if (toGradebook!=null && toGradebook.equals(EvaluationModelIfc.TO_DEFAULT_GRADEBOOK.toString()) &&
+    	if (toGradebook!=null && StringUtils.equals(toGradebook, String.valueOf(EvaluationModelIfc.TO_DEFAULT_GRADEBOOK)) &&
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/PublishAssessmentListener.java]: Same here

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 3]
```diff
@@ -124,6 +131,74 @@
...
+   * Check if there still exist a gradebook item which has the same name as the assessment AND is not externally maitained (empty).
+   * @param pTitle
+   * @return
+   */
+  public boolean isGradebookItemAvailable(String pTitle) {
+	  GradebookService g = (GradebookService) ComponentManager.get("org.sakaiproject.service.gradebook.GradebookService");
+	  String gradebookId = ToolManager.getInstance().getCurrentPlacement().getContext();
+	  List<Assignment> allGradebookItems = g.getAssignments(gradebookId);
+	  for(Assignment a : allGradebookItems) {
+		  if(!a.isExternallyMaintained() && StringUtils.equals(a.getName(), pTitle)) return true;
```
Comment (bjones86) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java]: Braces

**Comment by bjones86** [comment]
Some very minor comments, otherwise this LGTM. Extra points for cleaning up unnecessary boxing on some of those class members, replacing "magic numbers" with constants, and using StringUtils!

**Comment by fgeraci** [comment]
Hello @bjones86 , thank you for looking into this and your feedback. I made the suggested corrections and updated the commit.

**Comment by jonespm** [comment]
I think this is functionality that we want and the patch seems good to me. @fgeraci are you still around to resolve the conflict here? I can try to get time to test this locally but I feel like we should get this in because it's a better overall workflow.

**Comment by fgeraci** [comment]
Yes,  i will take a look at it either this afternoon or tomorrow.

Fg

**Comment by ern** [comment]
Conflicts need to be resolved

**Comment by fgeraci** [comment]
Thanks Earle. I am currently out of town, returning on Monday. I will
resume this then.

Regards

Fer Geraci

On Mar 2, 2017 8:31 AM, "Earle Nietzel" <notifications@github.com> wrote:

Assigned #2790 <https://github.com/sakaiproject/sakai/pull/2790> to @fgeraci
<https://github.com/fgeraci>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/sakaiproject/sakai/pull/2790#event-983958711>, or mute
the thread

**Comment by ncaidin** [comment]
@fgeraci , how is this one going?

**Comment by jonespm** [comment]
We do like this idea, but no comments for 6 months. This may need to be reviewed as part of a bigger gradebook consistency project, but I'd take it for 13 just for a short term fix too. Re-open if you resolve the conflicts.
